articles/tutorials/advanced/2d_shaders/06_color_swap_effect/index.md

```diff
@@ -0,0 +1,643 @@
+    float3 finalColor = lerp(grayscale, color.rgb, Saturation);
+
+    // Return the final color with the original alpha value
+    return float4(finalColor, color.a);
+}
+```
+
+In order for this to work, do not forget to add the `Saturation` shader parameter to the `colorSwapEffect.fx` file:
+
+```hlsl
+float Saturation;
+```
+
+For readability, extract the logic of the color swap effect into a new function as well:
+
+```hlsl
+float4 SwapColors(float4 color)
+{
+    // produce the key location
+    //  note the x-offset by half a texel solves rounding errors.
+    float2 keyUv = float2(color.r , 0);
+    
+    // read the swap color value
+    float4 swappedColor = tex2D(ColorMapSampler, keyUv) * color.a;
+    
+    // ignore the swap if the map does not have a value
+    bool hasSwapColor = swappedColor.a > 0;
+    if (!hasSwapColor)
+    {
+        return color;
+    }
+    
+    // return the result color
+    return lerp(swappedColor, color, OriginalAmount);
+}
+```
+
+And now the main shader function can chain these methods together:
+
+```hlsl
+float4 MainPS(VertexShaderOutput input) : COLOR
+{
+    // read the original color value
+    float4 originalColor = tex2D(SpriteTextureSampler,input.TextureCoordinates);
+
+    float4 swapped = SwapColors(originalColor);
+    float4 saturated = Grayscale(swapped);
+    
+    return saturated;
+}
+```
+
+> [!warning] 
+> Function Order Matters!
+> 
+> Make sure that the `Grayscale` and `SwapColors` functions appear _before_ the `MainPS` function in the shader, otherwise the compiler will not be able to resolve the functions.
+
+Now you can control the saturation manually with the debug slider,
+![Figure 6.18: Combining the color swap and saturation effect](./gifs/color-saturation.gif)
+
+The last thing to do is remove the old `grayscaleEffect` and re-write the game logic to set the `Saturation` parameter on the new effect. 
+In the `Draw()` method, instead of having an `if` case to start the `SpriteBatch` with different settings, it can always be configured to start with the `_colorSwapMaterial`:
+
+```csharp
+_colorSwapMaterial.SetParameter("Saturation", _saturation);
+Core.SpriteBatch.Begin(
+    samplerState: SamplerState.PointClamp,
+    sortMode: SpriteSortMode.Immediate,
+    effect: _colorSwapMaterial.Effect);
+```
+
+In the `Update()` method, we just need to set the `_saturation` back to `1` if the game is being played:
+
+```csharp
+if (_state != GameState.Playing)
+{
+    // The game is in either a paused or game over state, so
+    // gradually decrease the saturation to create the fading grayscale.
+    _saturation = Math.Max(0.0f, _saturation - FADE_SPEED);
+
+    // If its just a game over state, return back
+    if (_state == GameState.GameOver)
+    {
+        return;
+    }
+}
+else
+{
+    _saturation = 1;
+}
+```
+![Figure 6.19: The grayscale effect has been restored](./gifs/grayscale.gif)
+
+
+## Color Look Up Textures (LUTs)
```

Comment:
Good call outs to another technique, but please include links for "further reading", if the method is not being shown in this tutorial the readers shoudl know where to look for more info.  If it is being covered later, then include a chapter / link reference.

Don't drop a bomb and just leave them hanging :D